typescript/sdk/src/router/MultiProtocolRouterApps.ts

```diff
@@ -0,0 +1,85 @@
+import {
+  Address,
+  Domain,
+  ProtocolType,
+  promiseObjAll,
+} from '@hyperlane-xyz/utils';
+
+import { AdapterClassType, MultiProtocolApp } from '../app/MultiProtocolApp';
+import { ChainMap, ChainName } from '../types';
+
+import {
+  EvmGasRouterAdapter,
+  EvmRouterAdapter,
+} from './adapters/EvmRouterAdapter';
+import {
+  SealevelGasRouterAdapter,
+  SealevelRouterAdapter,
+} from './adapters/SealevelRouterAdapter';
+import { IGasRouterAdapter, IRouterAdapter } from './adapters/types';
+import { RouterAddress } from './types';
+
+export { Router } from '@hyperlane-xyz/core';
+
+export class MultiProtocolRouterApp<
+  ContractAddrs extends RouterAddress = RouterAddress,
+  IAdapterApi extends IRouterAdapter = IRouterAdapter,
+> extends MultiProtocolApp<ContractAddrs, IAdapterApi> {
+  // Explicit cast via unknown is unfortunately required here to enable both default adapters AND generic types
+  public readonly adapters = {
+    [ProtocolType.Ethereum]: EvmRouterAdapter as unknown as AdapterClassType<
+      ContractAddrs,
+      IAdapterApi
+    >,
+    [ProtocolType.Sealevel]:
+      SealevelRouterAdapter as unknown as AdapterClassType<
+        ContractAddrs,
+        IAdapterApi
+      >,
+  };
+
+  router(chain: ChainName): Address {
+    return this.metadata(chain).router;
+  }
+
+  interchainSecurityModules(): Promise<ChainMap<Address>> {
+    return promiseObjAll(
+      this.map((chain, adapter) => adapter.interchainSecurityModule(chain)),
+    );
+  }
+
+  owners(): Promise<ChainMap<Address>> {
+    return promiseObjAll(this.map((chain, adapter) => adapter.owner(chain)));
+  }
+
+  remoteRouters(
+    originChain: ChainName,
+  ): Promise<Array<{ domain: Domain; address: Address }>> {
+    return this.adapter(originChain).remoteRouters(originChain);
+  }
+}
+
+export class MultiProtocolGasRouterApp<
+  ContractAddrs extends RouterAddress = RouterAddress,
+  IAdapterApi extends IGasRouterAdapter = IGasRouterAdapter,
+> extends MultiProtocolRouterApp<ContractAddrs, IAdapterApi> {
+  // Explicit cast via unknown is unfortunately required here to enable both default adapters AND generic types
```

Comment:
are the defaults purely for backwards compatibility?

Reply:
Not in this case since this is new. It's for convenience of allowing use of MultiProtocolRouterApp without having to explicitly define those generic types.

Reply:
I was able to improve the typing around this, simplifying things a bit